composer/utils/sftp_object_store.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Utility for uploading to and downloading from cloud object stores."""
+import io
+import os
+import sys
+import uuid
+import warnings
+from typing import Iterator, Optional, Union
+
+from paramiko import SFTPAttributes, SSHClient
+
+from composer.utils.iter_helpers import IteratorFileStream
+
+__all__ = ["SFTPObjectStore"]
+
+
+class SFTPObjectStore:
+    """Utility for uploading to and downloading to a server via SFTP."""
+
+    def __init__(self, host: str, port: int = 22, username: Optional[str] = None, key_file_path: Optional[str] = None):
+        self.ssh_client = SSHClient()
+
+        # add AutoAddPolicy
+        if key_file_path is None:
+            self.ssh_client.load_system_host_keys()
+        else:
+            try:
+                self.ssh_client.load_system_host_keys(key_file_path)
+            except IOError:
+                warnings.warn(
+                    "Host keys could not be read from {key_file_path}. Attempting to read keys from known hosts.")
+                self.ssh_client.load_system_host_keys()
+
+        # implement some retries here
+        self.ssh_client.connect(host, port, username)
+
+        self.sftp_client = self.ssh_client.open_sftp()
+
+    def close_client(self):
+        self.ssh_client.close()
+
+    def upload_object(self, file_path: str, object_name: str):
+        """Upload an object currently located on a disk.
+
+        Args:
+            file_path (str): Path to the object on disk.
+            object_name (str): Object name (i.e. where the object will be stored in the container.)
+        """
+        # add verification and retries/rename
+        tmp_path = object_name + f".{uuid.uuid4()}.tmp"
+
+        object_size = os.stat(file_path).st_size
+        attributes = (SFTPAttributes)(self.sftp_client.put(file_path, tmp_path))
+        uploaded_bytes = attributes.st_size
+
+        if uploaded_bytes != object_size:
+            self.sftp_client.remove(tmp_path)
+            # Replace with custom exception later
+            raise Exception("File sizes don't match uploading error")
+        else:
+            self.sftp_client.rename(tmp_path, object_name)
+
+    def upload_object_via_stream(self, obj: Union[bytes, Iterator[bytes]], object_name: str):
```

Comment:
If `object_name` contains a subfolder path, do we need to manually create the nested directories if the folders do not yet exist?